VPlan: increase simplification power of simplifyRecipe

artagnon opened the pull request:
Use a worklist in simplifyRecipes to break up patterns in simplifyRecipe, increasing its simplification power.

-- 8< --
Based on #105699.

artagnon committed:
VPlan: use Worklist in simplifyRecipes

Since simplifyRecipe creates new recipes in some cases, use a Worklist
in its caller to capture newly-created recipes, and add it to the
Worklist, as a candidate for further simplification. This patch
thoroughly rewrites simplifyRecipe to simplify matched patterns,
eraseFromParent when applicable, and simplify the logic.

artagnon commented:
Rebase and ping.

artagnon commented:
Gentle ping.

artagnon commented:
Gentle ping. The objective of this patch should be straightforward.

artagnon commented:
Rebase and ping.

artagnon commented:
Gentle ping.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -986,83 +986,140 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
+  if (match(R, m_c_BinaryOr(m_VPValue(X), m_Not(m_VPValue(X1)))) && X == X1) {
+    auto *VPV = new VPValue(ConstantInt::getTrue(Ctx));
+    R->getVPSingleValue()->replaceAllUsesWith(VPV);
+    R->eraseFromParent();
```
There's nothing from preventing the same recipe being added multiple times to the worklist (probably at the moment it's not an issue, but can become one in the future). Better to let them be cleaned up later as is done now, to avoid this problem?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -986,83 +986,140 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
-  VPValue *A;
-  if (match(&R, m_Trunc(m_ZExtOrSExt(m_VPValue(A))))) {
-    VPValue *Trunc = R.getVPSingleValue();
+  VPValue *X, *X1, *Y, *Z;
```
Would be good to limit to changes adding the worklist and include other potential changes as NFC follow up? For this one, would be good to keep variables defined closer to their uses.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -986,83 +986,140 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
-static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+/// Try to simplify recipe \p R. Returns candidates for further simplification.
+static SmallVector<VPRecipeBase *>
+simplifyRecipe(VPRecipeBase *R, VPTypeAnalysis &TypeInfo, LLVMContext &Ctx) {
```
```suggestion
simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo, LLVMContext &Ctx) {
```

keep as &R to avoid unrelated changes?

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1046,23 +1048,73 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
         assert(TypeInfo.inferScalarType(VPV) == TypeInfo2.inferScalarType(VPV));
     }
 #endif
+    if (VPC)
+      return {VPC};
+    return {};
```
```suggestion
    return VPC ? {VPC} : {}
```

artagnon replied:
Unfortunately, the ternary operator in C/C++ is a special case, and this code is invalid: we can't put an initializer list in the RHS of a ternary operator.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -1046,23 +1048,73 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+  if (match(&R, m_BinaryOr(m_LogicalAnd(m_VPValue(X), m_VPValue(Y)),
+                           m_LogicalAnd(m_VPValue(X1), m_VPValue(Z)))) &&
+      X == X1) {
+    auto *YorZ = new VPInstruction(Instruction::Or, {Y, Z}, R.getDebugLoc());
```
Better to use VPBuilder to construct & Insert the new instructions?

fhahn commented on llvm/test/Transforms/LoopVectorize/SystemZ/pr47665.ll:
```diff
@@ -7,135 +7,109 @@ define void @test(ptr %p, i40 %a) {
-; CHECK-NEXT:    [[TMP5:%.*]] = icmp eq <16 x i1> [[TMP4]], zeroinitializer
-; CHECK-NEXT:    [[TMP6:%.*]] = icmp ult <16 x i1> zeroinitializer, [[TMP5]]
-; CHECK-NEXT:    [[TMP7:%.*]] = or <16 x i1> [[TMP6]], <i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true, i1 true>
-; CHECK-NEXT:    [[TMP8:%.*]] = icmp sgt <16 x i1> [[TMP7]], zeroinitializer
```
Do you know how this gets simplified? I might be missing something but, [[TMP7]] = 1, so should't `[[TMP8]]` be all-true? But below it's replaced by `false`?

artagnon replied:
So, I stepped through the code and didn't see anything surprising happening. The final VPlan, emitted as:

```
VPlan 'Final VPlan for VF={4},UF>=1' {
Live-in vp<%0> = VF * UF
Live-in vp<%1> = vector-trip-count
Live-in vp<%2> = backedge-taken count
Live-in ir<10> = original trip-count

vector.ph:
Successor(s): vector loop

<x1> vector loop: {
  vector.body:
    EMIT vp<%3> = CANONICAL-INDUCTION ir<0>, vp<%6>
    EMIT vp<%4> = WIDEN-CANONICAL-INDUCTION vp<%3>
    EMIT vp<%5> = icmp ule vp<%4>, vp<%2>
    WIDEN ir<%icmp.sgt> = icmp sgt ir<true>, ir<false>
  Successor(s): pred.store

  <xVFxUF> pred.store: {
    pred.store.entry:
      BRANCH-ON-MASK vp<%5>
    Successor(s): pred.store.if, pred.store.continue

    pred.store.if:
      REPLICATE store ir<%icmp.sgt>, ir<%p>
    Successor(s): pred.store.continue

    pred.store.continue:
    No successors
  }
  Successor(s): for.body.0

  for.body.0:
    EMIT vp<%6> = add vp<%3>, vp<%0>
    EMIT branch-on-count vp<%6>, vp<%1>
  No successors
}
Successor(s): middle.block

middle.block:
  EMIT branch-on-cond ir<true>
Successor(s): ir-bb<exit>, scalar.ph

ir-bb<exit>:
No successors

scalar.ph:
No successors
}
```

seems to be correct too. I'm confused about why `icmp sgt ir<true>, ir<false>` is lowered to `false`: is there an underlying bug somewhere (independent of this patch)?

artagnon replied:
Okay, I stepped through the ConstantFolder, and figured out that `icmp sgt true, false` is read as `-1 > 0`, since the APInt representation of true is `(1b, 1u, -1s)`, and we're asking for a signed comparison. I seriously doubt there's a bug so deep in the code, so I think the transformation is correct, and we just got confused.

fhahn replied:
Ah yes, the additional simplifications come from the IRBuilder + Folder, thanks for checking

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -984,23 +986,71 @@ static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
-      X == X1 && Y == Y1) {
+  // (X || true) -> true.
+  if (match(&R, m_c_BinaryOr(m_VPValue(X), m_True()))) {
+    auto *VPV = new VPValue(ConstantInt::getTrue(Ctx));
```
This should use `VPlan::getOrAddLiveIn`, otherwise this VPValue won't get freed properly

fhahn commented on llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:
```diff
@@ -892,8 +892,9 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
-static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+/// Try to simplify recipe \p R. Returns candidates for further simplification.
+static SmallVector<VPRecipeBase *>
+simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo, LLVMContext &Ctx) {
```
`Ctx` can be retrieved from `TypeInfo`

fhahn commented:
If it is not too much extra work, it would probably also be good to introduce the worklist as NFC commit without test changes and then replace the patterns.

artagnon commented:
> If it is not too much extra work, it would probably also be good to introduce the worklist as NFC commit without test changes and then replace the patterns.

Not sure what you mean by introducing the worklist as an NFC. Doesn't the introduction of worklist mean that we return candidates for further simplification, reducing recursively, and by definition, the change has higher simplification power? I might be able to introduce the worklist without any _test changes_, that isn't technically an NFC, with some difficulty; however, I don't see the value of such a patch.

fhahn commented:
> > If it is not too much extra work, it would probably also be good to introduce the worklist as NFC commit without test changes and then replace the patterns.
> 
> Not sure what you mean by introducing the worklist as an NFC. Doesn't the introduction of worklist mean that we return candidates for further simplification, reducing recursively, and by definition, the change has higher simplification power? I might be able to introduce the worklist without any _test changes_, that isn't technically an NFC, with some difficulty; however, I don't see the value of such a patch.

Ah Right, probably not entirely NFC, the main point would be to separate introducing the worklist from also adjusting the patterns, which should hopefully be fairly straigth-forward to separate as 2 distinct patches (the PR title already implies that the PR introduces a worklist only)

artagnon committed:
VPlan: introduce worklist in simplifyRecipes

In order to break up patterns in simplifyRecipes, and increase its
simplification power, introudce a worklist keeping a running list of
candidates for simplification, as a prelude to breaking up patterns in
simplifyRecipe.

artagnon commented:
> > > If it is not too much extra work, it would probably also be good to introduce the worklist as NFC commit without test changes and then replace the patterns.
> > 
> > 
> > Not sure what you mean by introducing the worklist as an NFC. Doesn't the introduction of worklist mean that we return candidates for further simplification, reducing recursively, and by definition, the change has higher simplification power? I might be able to introduce the worklist without any _test changes_, that isn't technically an NFC, with some difficulty; however, I don't see the value of such a patch.
> 
> Ah Right, probably not entirely NFC, the main point would be to separate introducing the worklist from also adjusting the patterns, which should hopefully be fairly straigth-forward to separate as 2 distinct patches (the PR title already implies that the PR introduces a worklist only)

Thanks, it was much simpler than I originally thought.

artagnon commented:
Honestly, I'm not sure what value #105699 adds, but this patch should be good now, either to be merged after #105699, or subsuming #105699 altogether.
